### ISSUE TITLE
Route withdraw is not sending the withdrawn route

### ISSUE THREAD

**Issue by adudek16** [Exa-Networks/exabgp#954]
##### Summary
[EDIT] When exabgp has not sent any routes out, eg when sending an anycast address for HA only when primary is down, route update does not contain the route being withdrawn, as result BGP session is torn down by neighbor with Missing Well-known Attribute.  Looking at the tcpdump differences, it appears that the route is not being sent the update. 

3.4.26 
```
        Update Message (2), length: 46
          Multi-Protocol Unreach NLRI (15), length: 20, Flags [O]:
            AFI: IPv6 (2), SAFI: Unicast (1)
              xxxx:xxxx:xxxx:xxxx::xxxx/128
```
vs
4.1.0
```
	Update Message (2), length: 30
	  Multi-Protocol Unreach NLRI (15), length: 3, Flags [OE]:
	    AFI: IPv6 (2), SAFI: Unicast (1)
	      End-of-Rib Marker (empty NLRI)
```

##### OS
Linux (Ubuntu 18.04.4)
```
Linux 4.15.0-51-generic #55-Ubuntu SMP Wed May 15 14:27:21 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```
##### Version
```
ExaBGP : 4.1.0-2074ac17
Python : 3.7.5 (default, Nov  7 2019, 10:50:52)  [GCC 8.3.0]
Uname  : Linux lab-1-de-nso 4.15.0-51-generic #55-Ubuntu SMP Wed May 15 14:27:21 UTC 2019 x86_64
Root   : /usr/local
```

##### Environment 

```
[exabgp.api]
encoder = text

[exabgp.daemon]
user = 'root'

[exabgp.log]
destination = '/ncs/log/exabgp/exabgp.log'
```

##### Configuration
```
neighbor 2003:1:ffe4:aa::1 {
        router-id 70.0.4.1;
        local-address 2003:1:ffe4:aa::2001;
        local-as 65550;
        peer-as 2792;
}

process announcer {
        run /usr/bin/python3 -m exabgp healthcheck --name master-check --ip 2003:1:ffff:aaaa::2001 --next-hop 2003:1:ffe4:aa::2001 --cmd "/get_ha_status.py -s" --no-syslog --withdraw-on-down;
        encoder json;
}

```

##### Importance
```
Minor as we are testing but the anycast announcement will be delayed until the session is established
```

**Comment by thomas-mangin** [comment]
related to https://github.com/Exa-Networks/exabgp/issues/573

**Comment by thomas-mangin** [comment]
start test
```
1> sudo ./qa/sbin/bgp --echo
2> ./sbin/exabgp etc/exabgp/conf-new-v6.conf -d
3> ./bin/exabgpcli s a o e
command: show adj-rib out extensive
neighbor 127.0.0.1 local-ip 127.0.0.1 local-as 65533 peer-as 65533 router-id 10.0.0.2 family-allowed in-open ipv6 unicast 2a02:b80:0:1::2/128 next-hop 2A02:B80:0:2::1 local-preference 200 community [ 30740:0 30740:30740 ]
neighbor 127.0.0.1 local-ip 127.0.0.1 local-as 65533 peer-as 65533 router-id 10.0.0.2 family-allowed in-open ipv6 unicast 2a02:b80:0:1::1/128 next-hop 2A02:B80:0:2::1 local-preference 200 community [ 30740:0 30740:30740 ]
```
then run either
```
3> ./bin/exabgpcli withdraw route 2a02:b80:0:1::2/128
3> ./bin/exabgpcli withdraw route 2a02:b80:0:1::2/128 next-hop 2A02:B80:0:2::1
```
generates
```
     127.0.0.1 52323 received 0: FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:003C:02:000000254001010040020040050400000064800F14000201802A020B80000000010000000000000002
     127.0.0.1 52323 sent     0: FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:003C:02:000000254001010040020040050400000064800F14000201802A020B80000000010000000000000002
```
decoded as
```
./sbin/exabgp etc/exabgp/conf-new-v6.conf -d --decode "0000 0025 4001 0100 4002 0040 0504 0000 0064 800F 1400 0201 802A 020B 8000 0000 0100 0000 0000 0000 02"
[...]
09:31:30 | 25019  | parser        | update json { "exabgp": "4.0.1", "time": 1581672690.387626, "host" : "ThomasMBP.local", "pid" : 25019, "ppid" : 19034, "counter": 1, "type": "update", "neighbor": { "address": { "local": "127.0.0.1", "peer": "127.0.0.1" }, "asn": { "local": 65533, "peer": 65533 } , "direction": "in", "message": { "update": { "attribute": { "origin": "igp", "local-preference": 100 }, "withdraw": { "ipv6 unicast": [ { "nlri": "2a02:b80:0:1::2/128" } ] } } } } }
```

QED? Update to a more recent version of ExaBGP ? I can not replicate on master.

**Comment by adudek16** [comment]
Installation: pip3 install -U exabgp==4.2.6
I am able to replicate this.  I will provide more data.

**Comment by thomas-mangin** [comment]
@adudek16 .. thank you as I can not fix what I can not replicate !

**Comment by thomas-mangin** [comment]
@adudek16 if you have both pip and apt/other package installed make sure you are running the latest version .. it caught quite a few users over the years :-p

**Comment by adudek16** [comment]
I think I have a simple way to reproduce this.
Simple bgp session established with no announcements.
```
root@exabgp-nso-test:/etc/exabgp# more exabgp.conf
neighbor 2a02:b80:fffb:11::1 {
        router-id 70.0.29.20;
        local-address 2a02:b80:fffb:11:8000::2020;
        local-as 65550;
        peer-as 65504;
}
```
Following your example:
```
root@exabgp-nso-test:/etc/exabgp# exabgp /etc/exabgp/exabgp.conf -d
22:58:40 | 16199  | welcome       | Thank you for using ExaBGP
22:58:40 | 16199  | version       | 4.2.6
22:58:40 | 16199  | interpreter   | 3.7.5 (default, Nov  7 2019, 10:50:52)  [GCC 8.3.0]
22:58:40 | 16199  | os            | Linux exabgp-nso-test 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:06:04 UTC 2019 x86_64
22:58:40 | 16199  | installation  | /usr/local
22:58:40 | 16199  | cli control   | named pipes for the cli are:
22:58:40 | 16199  | cli control   | to send commands  /run/exabgp.in
22:58:40 | 16199  | cli control   | to read responses /run/exabgp.out
22:58:40 | 16199  | configuration | performing reload of exabgp 4.2.6
22:58:40 | 16199  | configuration | > neighbor         | '2a02:b80:fffb:11::1'
22:58:40 | 16199  | configuration | . router-id        | '70.0.29.20'
22:58:40 | 16199  | configuration | . local-address    | '2a02:b80:fffb:11:8000::2020'
22:58:40 | 16199  | configuration | . local-as         | '65550'
22:58:40 | 16199  | configuration | . peer-as          | '65504'
22:58:40 | 16199  | configuration | < neighbor         |
22:58:40 | 16199  | reactor       | new peer: neighbor 2003:1b3b:fffb:11::1 local-ip 2003:1b3b:fffb:11:8000::2020 local-as 65550 peer-as 65504 router-id 70.0.29.20 family-allowed in-open
22:58:40 | 16199  | reactor       | loaded new configuration successfully
22:58:40 | 16199  | process       | forked process api-internal-cli-8a956aee
```
The bgp session comes up with the router as expected with no issues.   
Now I send 
`exabgpcli withdraw route 2a02:b80:fffa:ffff::2020/128 next-hop 2a02:b80:fffb:11:8000::2020` 

tcpdump capture shows
```
ffff ffff ffff ffff ffff ffff ffff ffff 0024 0200 0000 0d40 0101 0040 0206 0201 0001 000e
```
```
00:24:32 | 18433  | parser        | update json { "exabgp": "4.0.1", "time": 1581726272.4286563, "host" : "exabgp-nso-test", "pid" : 18433, "ppid" : 21825, "counter": 1, "type": "update", "neighbor": { "address": { "local": "2a02:b80:fffa:ffff::2020", "peer": "2a02:b80:fffb:11:8000::2020" }, "asn": { "local": 65550, "peer": 65504 } , "direction": "in", "message": { "update": { "attribute": { "origin": "igp", "as-path": [ 65550 ], "confederation-path": [] } } } } }
```

It appears that the update being generated is not compliant as it is missing the NEXT_HOP which is well-known mandatory for this update.
  
We can compare this behavior against 3.4.26 where the update is being sent as just a mp-unreach-nlri and the route is definitely in the withdraw message.  I may have chasing a red herring previously

**Comment by thomas-mangin** [comment]
did you ever announce the route first ? if not, this is probably where the difference lies and I can try that.

**Comment by adudek16** [comment]
No. We are using this to announce an anycast address only if the primary
server is down.

**Comment by thomas-mangin** [comment]
If you announce the IP before removing it the bug is not present .. I found an issue with the code bug fixing it is not trivial.

**Comment by adudek16** [comment]
Not just a missed logic case?

**Comment by thomas-mangin** [comment]
No more complex and the bug is hard to trigger as you must withdraw a route before announcing to trigger it

**Comment by adudek16** [comment]
I wouldn't say it is hard to trigger.  Do you understand the use case? I'm actually surprised that this hadn't been hit earlier since our use case is a standard way to do HA using an anycast address. I should check if the same thing happens if I use ipv4.   To me, it seems straight forward, we tell exabgp the exact route to withdraw, but the update is not generated with the information specified.  Previously, 3.4 behavior, it would send a Multi-Protocol Unreach NLRI with this withdraw. Now it looks like and IPv4 update instead. 
Was there a big refactor between 3.4 and 4.x? I only saw the one mentioned in the other issue.   Anything I can provide/help. I do admit that the addresses are not exact due to operational information, but I can provide data/information in a DM.

**Comment by adudek16** [comment]
Some more data points.
1) `exabgpcli withdraw route 10.10.100.100/32 next-hop 70.0.29.20`  
2a) `exabgpcli withdraw route 2003:1b3b:fffa:ffff::2020/128 next-hop 2003:1b3b:fffb:11:8000::2020`
or  
2b) `exabgpcli withdraw route 2003:1b3b:fffa:ffff::2020/128`
Then the issue does not appear, meaning the update is generated as expected 
```
ffff ffff ffff ffff  ffff ffff ffff ffff 003b 0200 0000 2440 0101 0040 0206 0201 0001 000e 800f 1400  0201 8020 031b 3bff faff ff00 0000 0000 0020 20
```
Not sure if this is relevant at the moment but 1) does not send any updates.

**Comment by thomas-mangin** [comment]
It is not hard to trigger, it is not triggered by normal use case. One should never send a withdrawal command if the announce was not sent previously.

**Comment by thomas-mangin** [comment]
Fixing the code correctly for when it is triggered is what is hard

**Comment by blny** [comment]
Hi, Thomas

Seem's I have also this problem. This my step for reproduce

1. Advertise route

Tue, 03 Mar 2020 11:42:02 189605 process       command from process ctrl-routes : announce route 2606:4700:3036::/64 next-hop ::ffff:195.34.52.115 community 8359:608
Tue, 03 Mar 2020 11:42:02 189605 reactor       async | ctrl-routes | announce route 2606:4700:3036::/64 next-hop ::ffff:195.34.52.115 community 8359:608
Tue, 03 Mar 2020 11:42:02 189605 api           route added to neighbor 172.16.1.121 local-ip 127.0.0.1 local-as 65499 peer-as 65499 router-id 10.0.0.1 family-allowed in-open : 2606:4700:3036
::/64 next-hop ::ffff:195.34.52.115 community 8359:608
Tue, 03 Mar 2020 11:42:02 189605 outgoing-1    sending TCP payload (  77) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 004D 0200 0000 3640 0101 0040 0200 4005 0400 0000 64C0 0804 20A7 0260 800E 1E00 0201 1000 0000 0000 0000 0000 00FF FFC3 2234 7300 4026 0647 0030 3600 00
Tue, 03 Mar 2020 11:42:02 189605 outgoing-1    >> 1 UPDATE(s)

12:00:35 | 190379 | parser        |
12:00:35 | 190379 | parser        | decoding routes in configuration
12:00:35 | 190379 | parser        | the message is an update
12:00:35 | 190379 | parser        | parsing UPDATE (  58) 0000 0036 4001 0100 4002 0040 0504 0000 0064 C008 0420 A702 6080 0E1E 0002 0110 0000 0000 0000 0000 0000 FFFF C322 3473 0040 2606 4700 3036 0000
12:00:35 | 190379 | routes        | withdrawn NLRI none
12:00:35 | 190379 | parser        | attribute origin             flag 0x40 type 0x01 len 0x01 payload 00
12:00:35 | 190379 | parser        | attribute as-path            flag 0x40 type 0x02 len 0x00
12:00:35 | 190379 | parser        | attribute local-preference   flag 0x40 type 0x05 len 0x04 payload 0000 0064
12:00:35 | 190379 | parser        | attribute community          flag 0xc0 type 0x08 len 0x04 payload 20A7 0260
12:00:35 | 190379 | parser        | attribute mp-reach-nlri      flag 0x80 type 0x0e len 0x1e payload 0002 0110 0000 0000 0000 0000 0000 FFFF C322 3473 0040 2606 4700 3036 0000
12:00:35 | 190379 | parser        | NLRI      ipv6 unicast       without path-information     payload 4026 0647 0030 3600 00
12:00:35 | 190379 | routes        | announced NLRI none
12:00:35 | 190379 | parser        |
12:00:35 | 190379 | parser        | decoded update 1 2606:4700:3036::/64 next-hop ::ffff:195.34.52.115 origin igp local-preference 100 community 8359:608
12:00:35 | 190379 | parser        | update json { "exabgp": "4.0.1", "time": 1583226035.6100147, "host" : "blny-macbook", "pid" : 190379, "ppid" : 16437, "counter": 1, "type": "update", "neighbor": { "address": { "local": "127.0.0.1", "peer": "172.16.1.121" }, "asn": { "local": 65499, "peer": 65499 } , "direction": "in", "message": { "update": { "attribute": { "origin": "igp", "local-preference": 100, "community": [ [ 8359, 608 ] ] }, "announce": { "ipv6 unicast": { "::ffff:195.34.52.115": [ { "nlri": "2606:4700:3036::/64" } ] } } } } } }

Check on quagga that route exist

blny-macbook# sh ipv6 route bgp
Codes: K - kernel route, C - connected, S - static, R - RIPng,
       O - OSPFv6, I - IS-IS, B - BGP, A - Babel, N - NHRP,
       > - selected route, * - FIB route

B>  2606:4700:3036::/64 [200/0] via ::ffff:195.34.52.115, net0.200 (recursive), 00:00:37
  *                               via fe80::213:7fff:fec1:2ad6, net0.200, 00:00:37

2. Withdraw route

Tue, 03 Mar 2020 11:43:54 189605 process       command from process ctrl-routes : withdraw route 2606:4700:3036::/64 next-hop ::ffff:195.34.52.115
Tue, 03 Mar 2020 11:43:54 189605 reactor       async | ctrl-routes | withdraw route 2606:4700:3036::/64 next-hop ::ffff:195.34.52.115
Tue, 03 Mar 2020 11:43:54 189605 api           route removed from neighbor 172.16.1.121 local-ip 127.0.0.1 local-as 65499 peer-as 65499 router-id 10.0.0.1 family-allowed in-open : 2606:4700:3036::/64 next-hop ::ffff:195.34.52.115
Tue, 03 Mar 2020 11:43:54 189605 outgoing-1    sending TCP payload (  52) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0034 0200 0000 1D40 0101 0040 0200 4005 0400 0000 6480 0F0C 0002 0140 2606 4700 3036 0000
Tue, 03 Mar 2020 11:43:54 189605 outgoing-1    >> 1 UPDATE(s)
Tue, 03 Mar 2020 11:43:54 189605 outgoing-1    received TCP payload (  19) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0016 03
Tue, 03 Mar 2020 11:43:54 189605 outgoing-1    received TCP payload (   3) 0303 03
Tue, 03 Mar 2020 11:43:54 189605 outgoing-1    << message of type NOTIFICATION
Tue, 03 Mar 2020 11:43:54 189605 outgoing-1    peer reset, message [notification received (3,3)] error[UPDATE message error / Missing Well-known Attribute / ^C]
Tue, 03 Mar 2020 11:43:54 189605 outgoing-1    outgoing-1 127.0.0.1-172.16.1.121, closing connection

12:02:05 | 190446 | parser        |
12:02:05 | 190446 | parser        | decoding routes in configuration
12:02:05 | 190446 | parser        | the message is an update
12:02:05 | 190446 | parser        | parsing UPDATE (  33) 0000 001D 4001 0100 4002 0040 0504 0000 0064 800F 0C00 0201 4026 0647 0030 3600 00
12:02:05 | 190446 | routes        | withdrawn NLRI none
12:02:05 | 190446 | parser        | attribute origin             flag 0x40 type 0x01 len 0x01 payload 00
12:02:05 | 190446 | parser        | attribute as-path            flag 0x40 type 0x02 len 0x00
12:02:05 | 190446 | parser        | attribute local-preference   flag 0x40 type 0x05 len 0x04 payload 0000 0064
12:02:05 | 190446 | parser        | attribute mp-unreach-nlri    flag 0x80 type 0x0f len 0x0c payload 0002 0140 2606 4700 3036 0000
12:02:05 | 190446 | parser        | NLRI      ipv6 unicast       without path-information     payload 4026 0647 0030 3600 00
12:02:05 | 190446 | routes        | announced NLRI none
12:02:05 | 190446 | parser        |
12:02:05 | 190446 | parser        | decoded update 2 2606:4700:3036::/64 origin igp local-preference 100
12:02:05 | 190446 | parser        | update json { "exabgp": "4.0.1", "time": 1583226125.4802787, "host" : "blny-macbook", "pid" : 190446, "ppid" : 16437, "counter": 1, "type": "update", "neighbor": { "address": { "local": "127.0.0.1", "peer": "172.16.1.121" }, "asn": { "local": 65499, "peer": 65499 } , "direction": "in", "message": { "update": { "attribute": { "origin": "igp", "local-preference": 100 }, "withdraw": { "ipv6 unicast": [ { "nlri": "2606:4700:3036::/64" } ] } } } } }

After that quagga drop bgp session, because of absent NH attribute

**Comment by adudek16** [comment]
I will double check but I believe that looks like the same behavior I see.    
@blny it looks like you are only doing IPv6 the same as me.

**Comment by blny** [comment]
@adudek16 I'm running IPv4 and IPv6 inside IPv4 session. With IPv4 all OK announce and withdraw route. With IPv6 in version 4.0.8 I don't see UNREACH NLRI at all. Upgrade exabgp to latest version 4.2.6 - I see UNREACH NLRI with removed prefix, but without NH.. So BGP session was dropped by BGP neighbor

**Comment by thomas-mangin** [comment]
@blny did you disable the adj-rib-out ? As if the Adj-Rib-Out is disabled that would explain why we can not get the next-hop from it if it is not provided on the command.

In any case, we should not attempt to send a invalid update and this should be raised to the user.

**Comment by blny** [comment]
Hi, @thomas-mangin  I checked both variant - adj-rib-out enabled and disabled.. result the same, then I send withdraw route with next-hop

Thu, 05 Mar 2020 12:09:46 231218 process       command from process ctrl-routes : withdraw route 2606:4700:3037::/64 next-hop ::ffff:195.34.52.115
 
outgoing BGP UPDATE missing NH in the mp-unreach-nlri

**Comment by subsecond** [comment]
Since @adudek16 pinged me regarding this issue here, I am trying to answer as best I can:

@adudek16 Did you update to ExaBGP 4.2.6? Can you still reproduce the initial behavior as described in https://github.com/Exa-Networks/exabgp/issues/954#issue-565002616? I was not able to reproduce your initial issue. Later in the thread you mention that only the nexthop attribute is missing (see my comment below) so I am not sure whether we are talking about two different things here.

@blny See https://github.com/Exa-Networks/exabgp/issues/573#issuecomment-614932392
TL;DR the nexthop attribute is a path attribute and according to [RFC4760](https://tools.ietf.org/html/rfc4760) it is not required when sending an `MP_UNREACH_NLRI` in an update message.
Basically, what you are facing is a bug in Quagga/FRR which is fixed by https://github.com/FRRouting/frr/commit/404c82d569a3150eeabce1c6480afdcc58ba97f1

I am confident that https://github.com/Exa-Networks/exabgp/issues/967 would fix the issue for BGP implementations that are either 
a) expecting only an `MP_UNREACH_NLRI` without any path attributes or 
b) that are parsing the path attributes and thereby expecting the "missing" nexthop attribute as well.

**Comment by adudek16** [comment]
@subsecond Yes. I can reproduce it easily with the steps above.

**Comment by subsecond** [comment]
Could you be a bit more specific, please:

Which part exactly can you reproduce? The missing next-hop attribute? Or is the entire withdrawn route missing? With ExaBGP 4.2.6 or 4.1.0?

The more details you provide the better one can understand what we are dealing with here.

**Comment by adudek16** [comment]
@subsecond 4.2.6 as stated on Feb 14 message. I will get fresh captures. today.

**Comment by thomas-mangin** [comment]
Could it be that your issue is that same as the one which was fixed with 4.2.10?

**Comment by adudek16** [comment]
@thomas-mangin  I can certainly verify.

**Comment by sincerywaing** [comment]
+1. the subprocess send route withdraw to main process, but the main process already closed pipe so main process does not receive withdraw from subprocess, then it won't send widthdraw to bgp neighbor. we saw this on 4.2.9 too. will take a look at 4.2.10.

**Comment by adudek16** [comment]
@thomas-mangin no, it does not.
@subsecond This happens in 4.2.10, 4.2.6, and 4.1.0. It is a behavior change from 3. 
What I see is exabgp sending an Update message (route withdraw) without a route.    
I see this on Alcatel Lucent 7x50 Routers running anything recent (16.0 - 19.10).

**Comment by calvincheng8** [comment]
The problem is fixed with the patch from https://github.com/thomas-mangin/exabgp/commit/1458cc325c1e9c9949272b6746eacb454fd4604b applied to 4.2.11. 

The patch was originally in #967 but never made it into any releases.

**Comment by adudek16** [comment]
@thomas-mangin 
@subsecond 
I can also say that with the patch #967 mentioned by @calvincheng8 the behavior is working as expected.
```
	Update Message (2), length: 46
	  Multi-Protocol Unreach NLRI (15), length: 20, Flags [O]:
	    AFI: IPv6 (2), SAFI: Unicast (1)
	      2003:1:ffff:8888::2001/128
```
Can we get that patch merged with 4.2.12?
